cpp/src/arrow/flight/internal.cc

```diff
@@ -49,27 +50,153 @@ Status FromGrpcStatus(const grpc::Status& grpc_status) {
+      break;
+    case grpc::StatusCode::ABORTED:
+      code = StatusCode::IOError;
+      message = util::StringBuilder("gRPC returned aborted error, with message: ",
+                                    grpc_status.error_message());
+      detail = std::make_shared<FlightStatusDetail>(FlightStatusCode::Internal);
+      break;
+    case grpc::StatusCode::OUT_OF_RANGE:
+      code = StatusCode::Invalid;
+      message = util::StringBuilder("gRPC returned out-of-range error, with message: ",
+                                    grpc_status.error_message());
+      break;
+    case grpc::StatusCode::UNIMPLEMENTED:
+      code = StatusCode::NotImplemented;
+      message = util::StringBuilder("gRPC returned unimplemented error, with message: ",
+                                    grpc_status.error_message());
+      break;
+    case grpc::StatusCode::INTERNAL:
+      code = StatusCode::IOError;
+      message = util::StringBuilder("gRPC returned internal error, with message: ",
+                                    grpc_status.error_message());
+      detail = std::make_shared<FlightStatusDetail>(FlightStatusCode::Internal);
+      break;
+    case grpc::StatusCode::UNAVAILABLE:
+      code = StatusCode::IOError;
+      message = util::StringBuilder("gRPC returned unavailable error, with message: ",
+                                    grpc_status.error_message());
+      detail = std::make_shared<FlightStatusDetail>(FlightStatusCode::Unavailable);
+      break;
+    case grpc::StatusCode::DATA_LOSS:
+      code = StatusCode::IOError;
+      message = util::StringBuilder("gRPC returned data loss error, with message: ",
+                                    grpc_status.error_message());
+      detail = std::make_shared<FlightStatusDetail>(FlightStatusCode::Internal);
+      break;
+    case grpc::StatusCode::UNAUTHENTICATED:
+      code = StatusCode::IOError;
+      message = util::StringBuilder("gRPC returned unauthenticated error, with message: ",
+                                    grpc_status.error_message());
+      detail = std::make_shared<FlightStatusDetail>(FlightStatusCode::Unauthenticated);
+      break;
+    default:
+      code = StatusCode::IOError;
+      message =
+          util::StringBuilder("gRPC failed with error code ", grpc_status.error_code(),
+                              " and message: ", grpc_status.error_message());
+      break;
   }
+  return Status(code, message, detail);
 }
 
 grpc::Status ToGrpcStatus(const Status& arrow_status) {
   if (arrow_status.ok()) {
     return grpc::Status::OK;
-  } else {
-    grpc::StatusCode grpc_code = grpc::StatusCode::UNKNOWN;
-    if (arrow_status.IsNotImplemented()) {
-      grpc_code = grpc::StatusCode::UNIMPLEMENTED;
-    } else if (arrow_status.IsInvalid()) {
-      grpc_code = grpc::StatusCode::INVALID_ARGUMENT;
+  }
+
+  grpc::StatusCode grpc_code = grpc::StatusCode::UNKNOWN;
+  std::string message = arrow_status.message();
```

Comment:
are you avoiding using Status.ToString() because it included Arrow's interpretation?

Reply:
Yes, otherwise the error gets a bit repetitive.